src/flamenco/runtime/program/fd_bpf_loader_v3_program.c

```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
     }
-
-    if ( instruction.inner.deploy_with_max_data_len.max_data_len < buffer_data_len ) {
+    if( instruction.inner.deploy_with_max_data_len.max_data_len<buffer_data_len ) {
+      FD_LOG_WARNING(( "Max data length is too small to hold Buffer data" ));
       return FD_EXECUTOR_INSTR_ERR_ACC_DATA_TOO_SMALL;
     }
-
-    if ( programdata_len > MAX_PERMITTED_DATA_LENGTH ) {
+    if( programdata_len>MAX_PERMITTED_DATA_LENGTH ) {
+      FD_LOG_WARNING(( "Max data length is too large" ));
       return FD_EXECUTOR_INSTR_ERR_INVALID_ARG;
     }
 
-    // let (derived_address, bump_seed) =
-    //             Pubkey::find_program_address(&[new_program_id.as_ref()], program_id);
-    //         if derived_address != programdata_key {
-    //             ic_logger_msg!(log_collector, "ProgramData address is not derived");
-    //             return Err(InstructionError::InvalidArgument);
-    //         }
-
-    // Drain buffer lamports to payer
-    int write_result = fd_instr_borrowed_account_modify( &ctx, buffer_acc, 0UL, &buffer_rec );
-    if( FD_UNLIKELY( write_result != FD_ACC_MGR_SUCCESS ) ) {
-      FD_LOG_WARNING(( "failed to read account metadata" ));
-      return FD_EXECUTOR_INSTR_ERR_MISSING_ACC;
+    /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L602-L608 */
+    /* Create ProgramData account */
+    fd_pubkey_t derived_address[ 1UL ];
+    uchar * seeds[ 1UL ];
+    seeds[ 0UL ] = (uchar *)new_program_id;
+    err = fd_pubkey_try_find_program_address( program_id, 1UL, seeds, derived_address );
+    if( err ) {
+      FD_LOG_WARNING(( "Failed to derive program address" ));
+      return err;
+    }
+    if( memcmp( derived_address, programdata_key, sizeof(fd_pubkey_t) ) ) {
+      FD_LOG_WARNING(( "ProgramData address is not derived" ));
+      return FD_EXECUTOR_INSTR_ERR_INVALID_ARG;
     }
 
+    /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L610-L627 */
+    /* Drain the Buffer account to payer before paying for programdata account */
     fd_borrowed_account_t * payer = NULL;
-    write_result = fd_instr_borrowed_account_modify( &ctx, payer_acc, 0UL, &payer );
-    if ( FD_UNLIKELY( write_result != FD_ACC_MGR_SUCCESS ) ) {
-      FD_LOG_WARNING(( "failed to read account metadata" ));
-      return FD_EXECUTOR_INSTR_ERR_MISSING_ACC;
-    }
-
-    /* Drain the buffer account to the payer before paying for the programdata account */
-    // FIXME: Do checked addition
-    ulong buffer_lamports = buffer_rec->meta->info.lamports;
-    payer->meta->info.lamports += buffer_lamports;
-    // TODO: Does this mean this account is dead?
-    buffer_rec->meta->info.lamports  = 0;
-
-    // TODO: deploy program
-    err = setup_program( &ctx, buffer_rec->data + BUFFER_METADATA_SIZE, buffer_data_len );
-    if ( err != 0 ) {
+    err = fd_instr_borrowed_account_modify_idx( instr_ctx, 0UL, 0UL, &payer );
```

Comment:
We should use `fd_borrowed_account_acquire_write`/`fd_borrowed_account_acquire_read`/`fd_borrowed_account_release_write`/`fd_borrowed_account_release_read` like we do in the other native programs.

We can move to the scoping macro in a follow-up PR.